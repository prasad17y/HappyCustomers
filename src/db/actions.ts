import {Q} from '@nozbe/watermelondb';
import {database} from '.';
import client from '../apollo/client';
import {LIST_ZELLER_CUSTOMERS} from '../apollo/queries';
import UserModel from './models/UserModel';
import {Role} from '../types/types';

// representation of the raw JSON data we receive from the GraphQL API
interface ApiCustomer {
  id: string;
  name: string | null;
  email: string | null;
  role: string | null;
}

export const syncData = async () => {
  try {
    // fetch data from the grphQL API
    const {data} = await client.query({
      query: LIST_ZELLER_CUSTOMERS,
    });

    // to simulate network delay
    await new Promise(resolve => setTimeout(resolve, 1000));

    const customersFromAPI: ApiCustomer[] = data.listZellerCustomers.items;

    // Filter out any invalid records from the API
    const validCustomers = customersFromAPI.filter(customer => {
      if (
        !customer.name ||
        !customer.role ||
        !Object.values(Role).includes(customer.role as Role)
      ) {
        console.log(`Filtering out invalid user from API: ID ${customer.id}`);
        return false;
      }
      return true;
    });

    await database.write(async () => {
      const usersCollection = database.collections.get<UserModel>('users');

      // array of ids of validCustomers
      const validCustomerIds = validCustomers.map(c => c.id);

      // fetch all users from db that are also present in validCustomers
      const localUsersFound = await usersCollection
        .query(Q.where('id', Q.oneOf(validCustomerIds)))
        .fetch();

      const localUsersById = new Map(localUsersFound.map(u => [u.id, u]));

      const operations = validCustomers.map(customer => {
        const localUser = localUsersById.get(customer.id);

        // update if customer already exists
        if (localUser) {
          return localUser.prepareUpdate(user => {
            user.name = customer.name!;
            user.email = customer.email;
            user.role = customer.role! as Role;
          });
        }

        return usersCollection.prepareCreate(user => {
          user._raw.id = customer.id;
          user.name = customer.name!;
          user.email = customer.email;
          user.role = customer.role! as Role;
        });
      });

      await database.batch(...operations);
      console.log('Database synchronized successfully (Upsert complete)!');
    });
  } catch (error) {
    console.log('Failed to sync data:', error);
    throw error;
  }
};

interface AddUserParams {
  firstName: string;
  lastName: string;
  email: string | null;
  role: Role;
}

export const addUser = async ({
  firstName,
  lastName,
  email,
  role,
}: AddUserParams) => {
  try {
    await database.write(async () => {
      const usersCollection = database.collections.get<UserModel>('users');
      await usersCollection.create(user => {
        // random Id is generated automatically
        user.name = `${firstName} ${lastName}`;
        user.email = email;
        user.role = role;
      });
    });
    console.log('User added successfully!');
  } catch (error) {
    console.log('Failed to add user:', error);
    throw error;
  }
};
