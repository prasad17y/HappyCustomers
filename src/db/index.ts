import {Database} from '@nozbe/watermelondb';
import SQLiteAdapter from '@nozbe/watermelondb/adapters/sqlite';

import {mySchema} from './schema';
import UserModel from './models/UserModel';

const adapter = new SQLiteAdapter({
  schema: mySchema,
});

export const database = new Database({
  adapter,
  modelClasses: [UserModel],
});
