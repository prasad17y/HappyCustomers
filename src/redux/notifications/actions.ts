import {
  SHOW_TOAST,
  CLEAR_TOAST,
  NotificationsActionTypes,
  ToastPayload,
} from './types';

export const showToast = (payload: ToastPayload): NotificationsActionTypes => ({
  type: SHOW_TOAST,
  payload,
});

export const clearToast = (): NotificationsActionTypes => ({
  type: CLEAR_TOAST,
});
