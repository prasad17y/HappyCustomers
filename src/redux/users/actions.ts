import {
  SYNC_USERS_REQUEST,
  SYNC_USERS_SUCCESS,
  SYNC_USERS_FAILURE,
  UsersActionTypes,
} from './types';

export const syncUsersRequest = (payload?: {
  forceRefresh: boolean;
}): UsersActionTypes => ({
  type: SYNC_USERS_REQUEST,
  payload,
});

export const syncUsersSuccess = (payload: {
  didFetch: boolean;
}): UsersActionTypes => ({
  type: SYNC_USERS_SUCCESS,
  payload,
});

export const syncUsersFailure = (error: string): UsersActionTypes => ({
  type: SYNC_USERS_FAILURE,
  payload: error,
});
