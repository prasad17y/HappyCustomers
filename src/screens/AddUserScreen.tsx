import React from 'react';
import {
  View,
  Text,
  TextInput,
  StyleSheet,
  TouchableOpacity,
  ScrollView,
  KeyboardAvoidingView,
  Platform,
  SafeAreaView,
  Animated,
  Keyboard,
} from 'react-native';
import {useNavigation} from '@react-navigation/native';
import {Role} from '../types/types';
import {addUser} from '../db/actions';
import {Formik} from 'formik';
import * as yup from 'yup';
import FilterTabs from '../components/FilterTabs';
import {ToastService} from '../services/ToastService';

export const validationSchema = yup.object().shape({
  firstName: yup
    .string()
    .trim()
    .matches(/^[a-zA-Z]*$/, 'Name can only contain letters.')
    .required('First Name is required.'),
  lastName: yup
    .string()
    .trim()
    .matches(/^[a-zA-Z]*$/, 'Name can only contain letters.')
    .required('Last Name is required.'),
  email: yup
    .string()
    .trim()
    .test(
      'is-valid-or-empty',
      'Please enter a valid email address.',
      (value: any) => {
        // .email() allows "john@doe", hence we need this elaborate test
        if (value === '') {
          return true;
        }
        return value ? /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(value) : true;
      },
    ),
  fullName: yup
    .string()
    .test(
      'max-length',
      'Full name cannot exceed 50 characters.',
      function (_value) {
        const {firstName = '', lastName = ''} = this.parent;
        return (firstName.trim() + ' ' + lastName.trim()).length <= 50;
      },
    ),
});

const AddUserScreen = () => {
  const navigation = useNavigation();
  const scrollPosition = React.useRef(new Animated.Value(1)).current;

  const handleSubmitCallback = async (values: {
    firstName: string;
    lastName: string;
    email: string;
    role: Role;
  }) => {
    try {
      await addUser({
        firstName: values.firstName.trim(),
        lastName: values.lastName.trim(),
        email: values.email.trim() || null,
        role: values.role,
      });
      ToastService.showSuccess('User created successfully!');
      navigation.goBack();
    } catch (error) {
      ToastService.showError('Could not create user.');
    }
  };

  return (
    <SafeAreaView style={styles.safeArea}>
      <KeyboardAvoidingView
        behavior={Platform.OS === 'ios' ? 'padding' : 'height'}
        style={styles.keyboardAvoiding}>
        <Formik
          initialValues={{
            firstName: '',
            lastName: '',
            email: '',
            role: Role.MANAGER,
            fullName: '',
          }}
          validationSchema={validationSchema}
          onSubmit={handleSubmitCallback}>
          {({
            handleChange,
            handleBlur,
            handleSubmit,
            values,
            errors,
            touched,
            setFieldValue,
          }) => (
            <ScrollView
              contentContainerStyle={styles.container}
              keyboardShouldPersistTaps="handled">
              <View style={styles.header}>
                <TouchableOpacity
                  onPress={() => navigation.goBack()}
                  style={styles.closeButton}>
                  <Text style={styles.closeButtonText}>âœ•</Text>
                </TouchableOpacity>
              </View>

              <Text style={styles.headerTitle}>New User</Text>

              <View style={styles.form}>
                <View style={styles.fieldContainer}>
                  <View style={styles.inputContainer}>
                    <TextInput
                      style={styles.input}
                      value={values.firstName}
                      onChangeText={handleChange('firstName')}
                      onBlur={handleBlur('firstName')}
                      placeholder="First Name"
                      placeholderTextColor="#a0a0a0"
                    />
                  </View>
                  <View style={styles.errorContainer}>
                    {errors.firstName && touched.firstName && (
                      <Text style={styles.errorText}>{errors.firstName}</Text>
                    )}
                    {errors.fullName && touched.firstName && (
                      <Text style={styles.errorText}>{errors.fullName}</Text>
                    )}
                  </View>
                </View>

                <View style={styles.fieldContainer}>
                  <View style={styles.inputContainer}>
                    <TextInput
                      style={styles.input}
                      value={values.lastName}
                      onChangeText={handleChange('lastName')}
                      onBlur={handleBlur('lastName')}
                      placeholder="Last Name"
                      placeholderTextColor="#a0a0a0"
                    />
                  </View>
                  <View style={styles.errorContainer}>
                    {errors.lastName && touched.lastName && (
                      <Text style={styles.errorText}>{errors.lastName}</Text>
                    )}
                    {errors.fullName && touched.lastName && (
                      <Text style={styles.errorText}>{errors.fullName}</Text>
                    )}
                  </View>
                </View>

                <View style={styles.fieldContainer}>
                  <View style={styles.inputContainer}>
                    <TextInput
                      style={styles.input}
                      value={values.email}
                      onChangeText={handleChange('email')}
                      onBlur={handleBlur('email')}
                      placeholder="Email"
                      placeholderTextColor="#a0a0a0"
                      keyboardType="email-address"
                      autoCapitalize="none"
                    />
                  </View>
                  <View style={styles.errorContainer}>
                    {errors.email && touched.email && (
                      <Text style={styles.errorText}>{errors.email}</Text>
                    )}
                  </View>
                </View>

                <Text style={styles.label}>User Role</Text>
                <FilterTabs
                  tabs={['Admin', 'Manager']}
                  selectedIndex={values.role === Role.ADMIN ? 0 : 1}
                  onTabPress={index => {
                    Keyboard.dismiss();
                    const role = index === 0 ? Role.ADMIN : Role.MANAGER;
                    setFieldValue('role', role);

                    Animated.spring(scrollPosition, {
                      toValue: index,
                      useNativeDriver: false,
                    }).start();
                  }}
                  scrollPosition={scrollPosition}
                  testID="add-user-screen"
                />
              </View>
              <View style={styles.footer}>
                <TouchableOpacity
                  style={styles.submitButton}
                  onPress={() => handleSubmit()}>
                  <Text style={styles.submitButtonText}>Create User</Text>
                </TouchableOpacity>
              </View>
            </ScrollView>
          )}
        </Formik>
      </KeyboardAvoidingView>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  safeArea: {
    flex: 1,
    backgroundColor: '#fff',
  },
  keyboardAvoiding: {
    flex: 1,
  },
  container: {
    flexGrow: 1,
    paddingHorizontal: 24,
  },
  header: {
    paddingTop: 16,
    alignItems: 'flex-start',
  },
  closeButton: {
    padding: 8,
  },
  closeButtonText: {
    fontSize: 24,
    color: '#0b5ac2',
  },
  headerTitle: {
    fontSize: 32,
    fontWeight: 'bold',
    color: '#191a1a',
    marginTop: 16,
    marginBottom: 32,
  },
  form: {
    flex: 1,
  },
  fieldContainer: {
    marginBottom: 8,
  },
  inputContainer: {
    borderBottomWidth: 1,
    borderBottomColor: '#e0e0e0',
  },
  input: {
    fontSize: 16,
    paddingVertical: 12,
    color: '#191a1a',
  },
  errorContainer: {
    height: 35,
    justifyContent: 'center',
  },
  errorText: {
    color: 'red',
  },
  label: {
    fontSize: 16,
    color: '#666',
    marginBottom: 12,
  },
  footer: {
    paddingVertical: 20,
  },
  submitButton: {
    backgroundColor: '#0b5ac2',
    padding: 16,
    borderRadius: 22,
    alignItems: 'center',
  },
  submitButtonText: {
    color: '#fff',
    fontSize: 18,
    fontWeight: 'bold',
  },
});

export default AddUserScreen;
