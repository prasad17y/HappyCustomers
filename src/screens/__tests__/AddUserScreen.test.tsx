import React from 'react';
import AddUserScreen, {validationSchema} from '../AddUserScreen';
import {addUser} from '../../db/actions';
import {ToastService} from '../../services/ToastService';
import {render, fireEvent, waitFor} from '@testing-library/react-native';

jest.mock('../../db/actions', () => ({
  addUser: jest.fn(),
}));

// Mock navigation
const mockGoBack = jest.fn();
jest.mock('@react-navigation/native', () => ({
  useNavigation: () => ({
    goBack: mockGoBack,
  }),
}));

// Mock ToastService
jest.mock('../../services/ToastService', () => ({
  ToastService: {
    showSuccess: jest.fn(),
    showError: jest.fn(),
  },
}));

describe('AddUserScreen Validation Schema', () => {
  it('should pass validation for correct data', async () => {
    const validData = {
      firstName: 'John',
      lastName: 'Doe',
      email: 'john.doe@example.com',
    };
    const isValid = await validationSchema.isValid(validData);
    expect(isValid).toBe(true);
  });

  it('should pass validation for an empty email string', async () => {
    const validData = {
      firstName: 'Jane',
      lastName: 'Doe',
      email: '',
    };
    const isValid = await validationSchema.isValid(validData);
    expect(isValid).toBe(true);
  });

  it('should fail validation if first name is missing', async () => {
    const invalidData = {firstName: '', lastName: 'Doe'};
    await expect(
      validationSchema.validate(invalidData, {abortEarly: true}),
    ).rejects.toThrow('First Name is required.');
  });

  it('should fail validation if name contains numbers', async () => {
    const invalidData = {firstName: 'John123', lastName: 'Doe'};
    await expect(validationSchema.validate(invalidData)).rejects.toThrow(
      'Name can only contain letters.',
    );
  });

  it('should fail validation for an invalid email format', async () => {
    const invalidData = {
      firstName: 'Jane',
      lastName: 'Doe',
      email: 'jane.doe@invalid',
    };
    await expect(validationSchema.validate(invalidData)).rejects.toThrow(
      'Please enter a valid email address.',
    );
  });

  it('should fail validation if full name is over 50 characters', async () => {
    const invalidData = {
      firstName: 'ThisIsAVeryLongFirstNameThatWillExceedTheLimit',
      lastName: 'ThisIsAVeryLongLastNameThatWillExceedTheLimit',
    };
    await expect(validationSchema.validate(invalidData)).rejects.toThrow(
      'Full name cannot exceed 50 characters.',
    );
  });
});

describe('AddUserScreen Component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly', () => {
    const {toJSON, getByTestId} = render(<AddUserScreen />);
    fireEvent(getByTestId('add-user-screen-filter-tabs-container'), 'layout', {
      nativeEvent: {layout: {width: 200}},
    });
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly with all validation errors visible', async () => {
    const {getByText, toJSON, getByTestId} = render(<AddUserScreen />);
    fireEvent(getByTestId('add-user-screen-filter-tabs-container'), 'layout', {
      nativeEvent: {layout: {width: 200}},
    });

    fireEvent.press(getByText('Create User'));

    await waitFor(() => {
      expect(getByText('First Name is required.')).toBeTruthy();
    });

    expect(toJSON()).toMatchSnapshot();
  });

  it('submits valid form and calls addUser & ToastService.showSuccess', async () => {
    (addUser as jest.Mock).mockResolvedValueOnce({});

    const {getByPlaceholderText, getByText} = render(<AddUserScreen />);

    fireEvent.changeText(getByPlaceholderText('First Name'), 'John');
    fireEvent.changeText(getByPlaceholderText('Last Name'), 'Doe');
    fireEvent.changeText(getByPlaceholderText('Email'), 'john.doe@example.com');

    fireEvent.press(getByText('Create User'));

    await waitFor(() => {
      expect(addUser).toHaveBeenCalledWith({
        firstName: 'John',
        lastName: 'Doe',
        email: 'john.doe@example.com',
        role: 'MANAGER',
      });
      expect(ToastService.showSuccess).toHaveBeenCalledWith(
        'User created successfully!',
      );
    });
  });

  it('submits valid form and calls addUser & ToastService.showSuccess with role admin', async () => {
    (addUser as jest.Mock).mockResolvedValueOnce({});

    const {getByPlaceholderText, getByText, getByTestId} = render(
      <AddUserScreen />,
    );

    fireEvent(getByTestId('add-user-screen-filter-tabs-container'), 'layout', {
      nativeEvent: {layout: {width: 200}},
    });

    const tab = getByTestId('add-user-screen-filter-tab-0'); // admin tab

    fireEvent.press(tab);

    fireEvent.changeText(getByPlaceholderText('First Name'), 'John');
    fireEvent.changeText(getByPlaceholderText('Last Name'), 'Doe');
    fireEvent.changeText(getByPlaceholderText('Email'), 'john.doe@example.com');

    fireEvent.press(getByText('Create User'));

    await waitFor(() => {
      expect(addUser).toHaveBeenCalledWith({
        firstName: 'John',
        lastName: 'Doe',
        email: 'john.doe@example.com',
        role: 'ADMIN',
      });
      expect(ToastService.showSuccess).toHaveBeenCalledWith(
        'User created successfully!',
      );
    });
  });

  it('shows error toast if addUser throws', async () => {
    (addUser as jest.Mock).mockRejectedValueOnce(
      new Error('Failed to add user'),
    );

    const {getByPlaceholderText, getByText} = render(<AddUserScreen />);

    fireEvent.changeText(getByPlaceholderText('First Name'), 'John');
    fireEvent.changeText(getByPlaceholderText('Last Name'), 'Doe');
    fireEvent.changeText(getByPlaceholderText('Email'), 'john.doe@example.com');

    fireEvent.press(getByText('Create User'));

    await waitFor(() => {
      expect(ToastService.showError).toHaveBeenCalledWith(
        'Could not create user.',
      );
    });
  });
});
