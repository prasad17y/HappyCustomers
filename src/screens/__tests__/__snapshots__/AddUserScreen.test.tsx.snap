// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddUserScreen Component renders correctly 1`] = `
<RCTSafeAreaView
  style={
    {
      "backgroundColor": "#fff",
      "flex": 1,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      [
        {
          "flex": 1,
        },
        {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <RCTScrollView
      contentContainerStyle={
        {
          "flexGrow": 1,
          "paddingHorizontal": 24,
        }
      }
      keyboardShouldPersistTaps="handled"
    >
      <View>
        <View
          style={
            {
              "alignItems": "flex-start",
              "paddingTop": 16,
            }
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
                "padding": 8,
              }
            }
          >
            <Text
              style={
                {
                  "color": "#0b5ac2",
                  "fontSize": 24,
                }
              }
            >
              âœ•
            </Text>
          </View>
        </View>
        <Text
          style={
            {
              "color": "#191a1a",
              "fontSize": 32,
              "fontWeight": "bold",
              "marginBottom": 32,
              "marginTop": 16,
            }
          }
        >
          New User
        </Text>
        <View
          style={
            {
              "flex": 1,
            }
          }
        >
          <View
            style={
              {
                "marginBottom": 8,
              }
            }
          >
            <View
              style={
                {
                  "borderBottomColor": "#e0e0e0",
                  "borderBottomWidth": 1,
                }
              }
            >
              <TextInput
                onBlur={[Function]}
                onChangeText={[Function]}
                placeholder="First Name"
                placeholderTextColor="#a0a0a0"
                style={
                  {
                    "color": "#191a1a",
                    "fontSize": 16,
                    "paddingVertical": 12,
                  }
                }
                value=""
              />
            </View>
            <View
              style={
                {
                  "height": 35,
                  "justifyContent": "center",
                }
              }
            />
          </View>
          <View
            style={
              {
                "marginBottom": 8,
              }
            }
          >
            <View
              style={
                {
                  "borderBottomColor": "#e0e0e0",
                  "borderBottomWidth": 1,
                }
              }
            >
              <TextInput
                onBlur={[Function]}
                onChangeText={[Function]}
                placeholder="Last Name"
                placeholderTextColor="#a0a0a0"
                style={
                  {
                    "color": "#191a1a",
                    "fontSize": 16,
                    "paddingVertical": 12,
                  }
                }
                value=""
              />
            </View>
            <View
              style={
                {
                  "height": 35,
                  "justifyContent": "center",
                }
              }
            />
          </View>
          <View
            style={
              {
                "marginBottom": 8,
              }
            }
          >
            <View
              style={
                {
                  "borderBottomColor": "#e0e0e0",
                  "borderBottomWidth": 1,
                }
              }
            >
              <TextInput
                autoCapitalize="none"
                keyboardType="email-address"
                onBlur={[Function]}
                onChangeText={[Function]}
                placeholder="Email"
                placeholderTextColor="#a0a0a0"
                style={
                  {
                    "color": "#191a1a",
                    "fontSize": 16,
                    "paddingVertical": 12,
                  }
                }
                value=""
              />
            </View>
            <View
              style={
                {
                  "height": 35,
                  "justifyContent": "center",
                }
              }
            />
          </View>
          <Text
            style={
              {
                "color": "#666",
                "fontSize": 16,
                "marginBottom": 12,
              }
            }
          >
            User Role
          </Text>
          <View
            onLayout={[Function]}
            style={
              [
                {
                  "backgroundColor": "#f0f2f5",
                  "borderRadius": 22,
                  "height": 44,
                  "padding": 4,
                },
                undefined,
              ]
            }
            testID="add-user-screen-filter-tabs-container"
          >
            <View
              style={
                {
                  "flex": 1,
                  "flexDirection": "row",
                  "position": "relative",
                }
              }
            >
              <View
                collapsable={false}
                style={
                  {
                    "backgroundColor": "#e3effa",
                    "borderColor": "#0b5ac1",
                    "borderRadius": 18,
                    "borderWidth": 1,
                    "height": "100%",
                    "position": "absolute",
                    "transform": [
                      {
                        "translateX": 96,
                      },
                    ],
                    "width": 96,
                  }
                }
              />
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "alignItems": "center",
                    "height": "100%",
                    "justifyContent": "center",
                    "opacity": 1,
                    "width": 96,
                  }
                }
                testID="add-user-screen-filter-tab-0"
              >
                <Text
                  style={
                    [
                      {
                        "color": "#666",
                        "fontSize": 16,
                        "fontWeight": "500",
                      },
                      false,
                    ]
                  }
                >
                  Admin
                </Text>
              </View>
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "alignItems": "center",
                    "height": "100%",
                    "justifyContent": "center",
                    "opacity": 1,
                    "width": 96,
                  }
                }
                testID="add-user-screen-filter-tab-1"
              >
                <Text
                  style={
                    [
                      {
                        "color": "#666",
                        "fontSize": 16,
                        "fontWeight": "500",
                      },
                      {
                        "color": "#0b5ac2",
                        "fontWeight": "bold",
                      },
                    ]
                  }
                >
                  Manager
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View
          style={
            {
              "paddingVertical": 20,
            }
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#0b5ac2",
                "borderRadius": 22,
                "opacity": 1,
                "padding": 16,
              }
            }
          >
            <Text
              style={
                {
                  "color": "#fff",
                  "fontSize": 18,
                  "fontWeight": "bold",
                }
              }
            >
              Create User
            </Text>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
</RCTSafeAreaView>
`;

exports[`AddUserScreen Component renders correctly with all validation errors visible 1`] = `
<RCTSafeAreaView
  style={
    {
      "backgroundColor": "#fff",
      "flex": 1,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      [
        {
          "flex": 1,
        },
        {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <RCTScrollView
      contentContainerStyle={
        {
          "flexGrow": 1,
          "paddingHorizontal": 24,
        }
      }
      keyboardShouldPersistTaps="handled"
    >
      <View>
        <View
          style={
            {
              "alignItems": "flex-start",
              "paddingTop": 16,
            }
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
                "padding": 8,
              }
            }
          >
            <Text
              style={
                {
                  "color": "#0b5ac2",
                  "fontSize": 24,
                }
              }
            >
              âœ•
            </Text>
          </View>
        </View>
        <Text
          style={
            {
              "color": "#191a1a",
              "fontSize": 32,
              "fontWeight": "bold",
              "marginBottom": 32,
              "marginTop": 16,
            }
          }
        >
          New User
        </Text>
        <View
          style={
            {
              "flex": 1,
            }
          }
        >
          <View
            style={
              {
                "marginBottom": 8,
              }
            }
          >
            <View
              style={
                {
                  "borderBottomColor": "#e0e0e0",
                  "borderBottomWidth": 1,
                }
              }
            >
              <TextInput
                onBlur={[Function]}
                onChangeText={[Function]}
                placeholder="First Name"
                placeholderTextColor="#a0a0a0"
                style={
                  {
                    "color": "#191a1a",
                    "fontSize": 16,
                    "paddingVertical": 12,
                  }
                }
                value=""
              />
            </View>
            <View
              style={
                {
                  "height": 35,
                  "justifyContent": "center",
                }
              }
            >
              <Text
                style={
                  {
                    "color": "red",
                  }
                }
              >
                First Name is required.
              </Text>
            </View>
          </View>
          <View
            style={
              {
                "marginBottom": 8,
              }
            }
          >
            <View
              style={
                {
                  "borderBottomColor": "#e0e0e0",
                  "borderBottomWidth": 1,
                }
              }
            >
              <TextInput
                onBlur={[Function]}
                onChangeText={[Function]}
                placeholder="Last Name"
                placeholderTextColor="#a0a0a0"
                style={
                  {
                    "color": "#191a1a",
                    "fontSize": 16,
                    "paddingVertical": 12,
                  }
                }
                value=""
              />
            </View>
            <View
              style={
                {
                  "height": 35,
                  "justifyContent": "center",
                }
              }
            >
              <Text
                style={
                  {
                    "color": "red",
                  }
                }
              >
                Last Name is required.
              </Text>
            </View>
          </View>
          <View
            style={
              {
                "marginBottom": 8,
              }
            }
          >
            <View
              style={
                {
                  "borderBottomColor": "#e0e0e0",
                  "borderBottomWidth": 1,
                }
              }
            >
              <TextInput
                autoCapitalize="none"
                keyboardType="email-address"
                onBlur={[Function]}
                onChangeText={[Function]}
                placeholder="Email"
                placeholderTextColor="#a0a0a0"
                style={
                  {
                    "color": "#191a1a",
                    "fontSize": 16,
                    "paddingVertical": 12,
                  }
                }
                value=""
              />
            </View>
            <View
              style={
                {
                  "height": 35,
                  "justifyContent": "center",
                }
              }
            />
          </View>
          <Text
            style={
              {
                "color": "#666",
                "fontSize": 16,
                "marginBottom": 12,
              }
            }
          >
            User Role
          </Text>
          <View
            onLayout={[Function]}
            style={
              [
                {
                  "backgroundColor": "#f0f2f5",
                  "borderRadius": 22,
                  "height": 44,
                  "padding": 4,
                },
                undefined,
              ]
            }
            testID="add-user-screen-filter-tabs-container"
          >
            <View
              style={
                {
                  "flex": 1,
                  "flexDirection": "row",
                  "position": "relative",
                }
              }
            >
              <View
                collapsable={false}
                style={
                  {
                    "backgroundColor": "#e3effa",
                    "borderColor": "#0b5ac1",
                    "borderRadius": 18,
                    "borderWidth": 1,
                    "height": "100%",
                    "position": "absolute",
                    "transform": [
                      {
                        "translateX": 96,
                      },
                    ],
                    "width": 96,
                  }
                }
              />
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "alignItems": "center",
                    "height": "100%",
                    "justifyContent": "center",
                    "opacity": 1,
                    "width": 96,
                  }
                }
                testID="add-user-screen-filter-tab-0"
              >
                <Text
                  style={
                    [
                      {
                        "color": "#666",
                        "fontSize": 16,
                        "fontWeight": "500",
                      },
                      false,
                    ]
                  }
                >
                  Admin
                </Text>
              </View>
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "alignItems": "center",
                    "height": "100%",
                    "justifyContent": "center",
                    "opacity": 1,
                    "width": 96,
                  }
                }
                testID="add-user-screen-filter-tab-1"
              >
                <Text
                  style={
                    [
                      {
                        "color": "#666",
                        "fontSize": 16,
                        "fontWeight": "500",
                      },
                      {
                        "color": "#0b5ac2",
                        "fontWeight": "bold",
                      },
                    ]
                  }
                >
                  Manager
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View
          style={
            {
              "paddingVertical": 20,
            }
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#0b5ac2",
                "borderRadius": 22,
                "opacity": 1,
                "padding": 16,
              }
            }
          >
            <Text
              style={
                {
                  "color": "#fff",
                  "fontSize": 18,
                  "fontWeight": "bold",
                }
              }
            >
              Create User
            </Text>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
</RCTSafeAreaView>
`;
